Fix floating-point comparison in QueryAssertions for aggregation

kagamiori opened the pull request:
Summary: assertEqualResults and assertResults in QueryAssertions.cpp compare the actual and expected result vectors through a data structure std::multiset<MaterializedRow, MaterializedRowComparator> where floating-point values are compared with epsilon. This data structure has undefined behavior because MaterializedRowComparator doesn't satisfy the strict weak ordering required by std (https://github.com/facebookincubator/velox/issues/3493). This bug in particular affects the verification of aggregation functions because floating-point results are computed via different query plans and hence the comparison needs to tolerate imprecision. This diff fixes this bug for aggregation by reordering the result columns to put the floating-point column at the end and grouping the results via sorting by the rest columns. If there is only one floating-point column or every group only contains one result row, the result rows are then compared via a merge-join logic with epsilon. Otherwise, the result rows are compared directly without epsilon.

The requirement for every group containing only one result row when there are multiple floating point columns is necessary because the imprecision in one floating point column can make the sorted rows disordered with regards to the other floating point columns.
E.g., Expected rows:
| c0 | c1 |
|-----|------|
| 1.0 | 1.0 |
| 1.0001 | 2.0 |

Actual rows:
| c0 | c1 |
|----|----|
|1.0001 | 2.0 |
| 1.001 | 1.0 |

Window aggregation test is tricky because there can be multiple result rows for one group but it requires comparison with epsilon. So we add an additional identifier input vector to force each group in the result containing one row only.
E.g., (c0 for input, p0 for partition key, w0 for window function result, id for the additional identifier column.)
| c0 | p0 | w0 | id |
|---|---|---|---|
| 1.0 | 1 | 0.123 | 0 |
| 1.1 | 1 | 0.234 | 1 |

Reviewed By: Yuhta

Differential Revision: D42557674

mbasmanova commented on velox/vector/fuzzer/VectorFuzzer.h:
```diff
@@ -223,6 +223,10 @@ class VectorFuzzer {
 
+  // Generate a random non-floating-point primitive type to be used as group key
+  // for aggregations.
+  TypePtr randGroupKeyType();
```
typos: "group key" -> "group-by key" or "grouping key". This applies to other keys as well, e.g. join key. Maybe, rename to avoid specifying the usage of the key: randScalarNonFloatingPointType.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/SumTest.cpp:
```diff
@@ -223,7 +223,7 @@ TEST_F(SumTest, sumFloat) {
 }
 
-TEST_F(SumTest, sumDoubleAndFloatg) {
+TEST_F(SumTest, sumDoubleAndFloat) {
```
Thank you for fixing this typo.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.h:
```diff
@@ -44,9 +44,8 @@ struct MaterializedRowComparator {
-using MaterializedRowMultiset =
-    std::multiset<MaterializedRow, MaterializedRowComparator>;
+// Multiset that compares floating-point values directly.
+using MaterializedRowMultiset = std::multiset<MaterializedRow>;
```
As you found out the MaterializedRowComparator cannot be used as a comparator. Let's remove it to avoid confusion in the future.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -719,6 +894,14 @@ static bool compareMaterializedRows(
     }
   }
 
+  // In case there are values in right but not in left, check the other way
```
Perhaps, clarify that this applies to cases where there are duplicate values in 'left'. If all values in 'left' are unique, this additional loop is not necessary.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -719,6 +894,14 @@ static bool compareMaterializedRows(
 
+  // In case there are values in right but not in left, check the other way
+  // around. E.g., left = {1, 1, 2}, right = {1, 2, 3}.
+  for (auto& it : right) {
```
This is a nice catch, but seems separate from the other issue you are fixing. It would be nice to extract it into a separate PR.

kagamiori replied:
Yeah, this is fixed in https://github.com/facebookincubator/velox/pull/3725.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -763,11 +945,20 @@ static bool compareOrderedPartitions(
-  }
 
-  return true;
+  std::vector<MaterializedRow> leftReordered{
```
It seems like tryBucketMaterializedRows can be called with empty xxxReordered. The function can fill these in on its own. This will make the calling code simpler.

mbasmanova reviewed (commented):
@kagamiori Wei, thank you for working on this problem. Some comments and questions below. Should we add a test to verify the new logic?

kagamiori commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +780,107 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  auto width = leftReordered[0].size();
+  auto length = leftReordered.size();
+
+  // Compare row-by-row with epsilon.
```
I thought about this, but was wondering whether there will be rows with different width in one MaterializedRowMultiset in practice, since it's either returned from DuckDB execution or converted from a RowVector. I can add this check for every row.

kagamiori commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +780,107 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// Try to bucket rows in left and right into leftReordered and rightReordered
+// and returns whether the bucketization is successful. The bucketization is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
```
For use cases of this tryBucketMaterializedRows() functions, one of left and right is the expected result and the other is the actual result computed by our library. It's possible that the expected and actual results do not match in terms of the number of rows or columns. In this case, we don't try to bucket the results.

kagamiori commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +780,107 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+    }
+  }
+
+  auto moveFloatColumnsToEnd =
```
This works correct if floatColumnIndices is empty, but in this situation it's unnecessary to attempt to do the transformation. I'll add a check for floatColumnIndices.empty() and return false in this case.

kagamiori commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +780,107 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  std::sort(leftReordered.begin(), leftReordered.end());
+  std::sort(rightReordered.begin(), rightReordered.end());
+
+  // Check that every bucket grouped by the first width - 1 columns has only
```
Typo. I'll update it to "Check that every bucket grouped by non-floating-point columns has only one row".

mbasmanova commented on velox/exec/tests/AggregationFuzzer.cpp:
```diff
@@ -152,6 +153,14 @@ class AggregationFuzzer {
+  // child named "id" of BIGINT identifiers that differentiates every row.
+  // Identifier starts from 0. This additional input vector is needed for
+  // result verification of window aggregations.
+  std::vector<RowVectorPtr> generateInputDataWithId(
```
nit: here ID is really a row number, hence, consider renaming to generateInputDataWithRowNumber and use "row_number" as column name

mbasmanova commented on velox/exec/tests/AggregationFuzzer.cpp:
```diff
@@ -491,6 +500,31 @@ std::vector<RowVectorPtr> AggregationFuzzer::generateInputData(
+
+  std::vector<RowVectorPtr> input;
+  auto size = vectorFuzzer_.getOptions().vectorSize;
+  int64_t identifier = 0;
```
identifier -> rowNumber

mbasmanova commented on velox/exec/tests/AggregationFuzzer.cpp:
```diff
@@ -491,6 +500,31 @@ std::vector<RowVectorPtr> AggregationFuzzer::generateInputData(
+  int64_t identifier = 0;
+  for (auto i = 0; i < FLAGS_num_batches; ++i) {
+    std::vector<VectorPtr> children;
+    for (auto i = 0; i + 1 < inputType->size(); ++i) {
```
nit: i < inputType->size() - 1 for readability

kagamiori replied:
"+1" on the left side is actually needed to avoid unsigned integer underflow because inputType->size() is a uint32_t. (inputType->size() usually should not be 0 though, but just in case.)

mbasmanova replied:
Can we add VELOX_CHECK_GT(inputType->size(), 0) to be sure ?

kagamiori replied:
After the change, I found that we can be sure that types.size() is at least one because we manually inserted BIGINT() to it. So I just moved "+1" on the left side to "-1" on the right side.

mbasmanova commented on velox/exec/tests/AggregationFuzzer.cpp:
```diff
@@ -491,6 +500,31 @@ std::vector<RowVectorPtr> AggregationFuzzer::generateInputData(
+    for (auto i = 0; i + 1 < inputType->size(); ++i) {
+      children.push_back(vectorFuzzer_.fuzz(inputType->childAt(i), size));
+    }
+    velox::test::VectorMaker vectorMaker{pool_.get()};
```
Do we need a separate instance for VectorMaker for each batch? Perhaps, use a single instance defined before the loop.

mbasmanova commented on velox/exec/tests/AggregationFuzzer.cpp:
```diff
@@ -491,6 +500,31 @@ std::vector<RowVectorPtr> AggregationFuzzer::generateInputData(
+    velox::test::VectorMaker vectorMaker{pool_.get()};
+    children.push_back(vectorMaker.flatVector<int64_t>(
+        size, [&](auto /*row*/) { return identifier++; }));
+    auto augmentedInput = std::make_shared<RowVector>(
```
It is confusing that input and augmentedInput variables are of different types (one is an std::vector, another a VectorPtr). Consider dropping augmentedInput variable altogether.

```
input.push_back(vectorMaker.rowVector(children));
```

mbasmanova reviewed (commented):
@kagamiori Wei, thank you for working on this. I still need to read the code changes QueryAssertions.cpp, but here are some initial comments on other parts of the PR.

> Window aggregation test is tricky because if its input is a floating-point column, result rows grouped by non-floating-point columns won't satisfy the requirement of bucket size being 1. So we project out the input columns after getting the window aggregation result in this situation.

This part of the PR description is not very clear and doesn't seem to match the code changes. Specifically, it is not clear how "we project out the input columns after getting the window aggregation result" helps fix the issue. From the code changes I'm seeing that you are adding an extra "row number" column to the input with unique values of type BIGINT. 

It seems that we are introducing a limitation to assertQuery and similar methods, i.e. inputs with floating points columns can be compared only if non-floating point columns form a unique key. It would be nice to document that limitation.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,6 +550,95 @@ std::string toString(const MaterializedRow& row) {
+         actualIndex < actualRows.size()) {
+    bool equal = true;
+    if (equalWithEpsilon(
+            expectedRows[expectedIndex], actualRows[actualIndex])) {
```
nit: consider extracting temp variables for expectedRows[expectedIndex] and actualRows[actualIndex] to reduce repetition and improve readability.

```
    const auto& expectedRow = expectedRows[expectedIndex];
    const auto& actualRow = actualRows[actualIndex];
    if (equalWithEpsilon(expectedRow, actualRow)) {
      ++expectedIndex;
      ++actualIndex;
    } else if (lessThanWithEpsilon(expectedRow, actualRow)) {
      missingActualRows.push_back(expectedRow);
      ++expectedIndex;
    } else {
      extraActualRows.push_back(actualRow);
      ++actualIndex;
    }
```

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,6 +550,95 @@ std::string toString(const MaterializedRow& row) {
+      ++actualIndex;
+    }
+  }
+  while (actualIndex < actualRows.size()) {
```
nit:

```
for (; actualIndex < actualRows.size(); ++actualIndex) {
   extraActualRows.push_back(actualRows[actualIndex]);
}
```

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,6 +550,95 @@ std::string toString(const MaterializedRow& row) {
+    extraActualRows.push_back(actualRows[actualIndex]);
+    ++actualIndex;
+  }
+  while (expectedIndex < expectedRows.size()) {
```
same

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,6 +550,95 @@ std::string toString(const MaterializedRow& row) {
+    ++expectedIndex;
+  }
+
+  std::ostringstream message;
```
This logic seems to be duplicated. Consider extracting it into a helper function and reuse.

mbasmanova reviewed (commented):
@kagamiori Overall looks good. Some comments below.

kagamiori commented:
> @kagamiori Wei, thank you for working on this. I still need to read the code changes QueryAssertions.cpp, but here are some initial comments on other parts of the PR.
> 
> > Window aggregation test is tricky because if its input is a floating-point column, result rows grouped by non-floating-point columns won't satisfy the requirement of bucket size being 1. So we project out the input columns after getting the window aggregation result in this situation.
> 
> This part of the PR description is not very clear and doesn't seem to match the code changes. Specifically, it is not clear how "we project out the input columns after getting the window aggregation result" helps fix the issue. From the code changes I'm seeing that you are adding an extra "row number" column to the input with unique values of type BIGINT.
> 
> It seems that we are introducing a limitation to assertQuery and similar methods, i.e. inputs with floating points columns can be compared only if non-floating point columns form a unique key. It would be nice to document that limitation.

Hi Masha, the PR description was for the last version of the code. I updated it internally but it didn't get synced up. 

> @kagamiori Wei, thank you for working on this. I still need to read the code changes QueryAssertions.cpp, but here are some initial comments on other parts of the PR.
> 
> > Window aggregation test is tricky because if its input is a floating-point column, result rows grouped by non-floating-point columns won't satisfy the requirement of bucket size being 1. So we project out the input columns after getting the window aggregation result in this situation.
> 
> This part of the PR description is not very clear and doesn't seem to match the code changes. Specifically, it is not clear how "we project out the input columns after getting the window aggregation result" helps fix the issue. From the code changes I'm seeing that you are adding an extra "row number" column to the input with unique values of type BIGINT.
> 
> It seems that we are introducing a limitation to assertQuery and similar methods, i.e. inputs with floating points columns can be compared only if non-floating point columns form a unique key. It would be nice to document that limitation.

Hi Masha, the PR description was for the last version of the code. I updated the code and description internally, but the description didn't get synced to GitHub. I've updated it manually and will add comments to the code as well.

kagamiori commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +793,115 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// most one row.
+// The bucket results are stored in leftBucketed and rightBucketed with the
+// floating-point columns moved to the end and rows in ascending order.
+bool tryBucketMaterializedRows(
```
This function groups rows by non-floating-point columns via sorting and if there are multiple floating-point columns, it checks that every group contains at most one row. (I feel the other name I think of is worse than the current one, i.e., "qualifyEpsilonComparison".)

mbasmanova replied:
> if there are multiple floating-point columns, it checks that every group contains at most one row

Does the check apply only when there are multiple floating point columns. I assume it applies if there is at least one such column, no?

> I feel the other name I think of is worse

Let's try to come up with a few names. Maybe, sortByUniqueKey(<list of columns that are expected to form a unique key>).

kagamiori replied:
> Does the check apply only when there are multiple floating point columns.
I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.

> Maybe, sortByUniqueKey().
This sounds like we already have unique keys and then sort by them. But this function is more like we sort the data to check whether all keys are unique.

mbasmanova replied:
> I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.

I'm curious why. It seems it doesn't matter how many floating-point column there are. What's the difference between having 1 or 2 or more?

mbasmanova replied:
> I feel this sounds like we already have unique keys and then sort by them. But this function is more like we sort the data to check whether all keys are unique.

Another way to look at this is that we expect that a subset of columns forms a unique key and we want to sort by them. In case our assumption is wrong this function will return 'false'.

kagamiori replied:
> > I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.
> 
> I'm curious why. It seems it doesn't matter how many floating-point column there are. What's the difference between having 1 or 2 or more?

We can check every group containing a single row regardless of how many floating-point columns we have. It's just when there is one floating-point column, even if there are multiple rows per group, we can still compare them via sort-merge with epsilon (because even if there are two expected values x < y that are close enough such that their actual approximate x' > y', y' and x' can still match with x and y.) 

If this case makes it hard to understand the behavior of assertEqualResults(), I can just force it to check every group containing one row for the case of one floating-point column too.

kagamiori replied:
According to offline discussion, I removed the optimization for the case of one floating-point column. I added a comment about relaxing the condition of single-row-groups under this case in QueryAssertions.h.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,28 +550,41 @@ std::string toString(const MaterializedRow& row) {
+}
 
+std::string makeErrorMessage(
+    const std::vector<MaterializedRow>& missingActualRows,
```
nit: consider renaming to missingRows and actualRows. missingActualRows doesn't quite make sense, because these rows are missing from the 'actual' results.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -595,6 +609,71 @@ std::string generateUserFriendlyDiff(
+    missingActualRows.push_back(expectedRows[expectedIndex]);
+  }
+
+  return makeErrorMessage(
```
Nice refactor.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  }
+}
+
+// Return true if currentRow and previsouRow are the same at the first width
```
typo: previsouRow

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+    const MaterializedRow& currentRow,
+    const MaterializedRow& previousRow,
+    size_t width) {
+  for (auto j = 0; j < width; ++j) {
```
nit: j -> i

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// width columns. The argument groupedBuckets should be sorted by the first
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& groupedBuckets,
```
nit: groupedBuckets -> sortedRows ?

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  std::sort(leftBucketed.begin(), leftBucketed.end());
+  std::sort(rightBucketed.begin(), rightBucketed.end());
+
+  if (floatColumnIndices.size() <= 1) {
```
Why is the case of single floating-point column is special?

kagamiori replied:
According to offline discussion, I removed this special case for floatColumnIndices.size() <= 1.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+    return false;
+  }
+
+  std::vector<velox::column_index_t> floatColumnIndices;
```
I still feel that it would be more readable to extract the logic of computing the list of floating-point column indices into a separate function.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  // Check that every bucket grouped by non-floating-point columns has only
+  // one row.
+  width = width - floatColumnIndices.size();
+  bool leftOversized = hasMultiRowGroups(leftBucketed, width);
```
The usage of term "oversized" here is confusing to me. Would it be possible to rename or just remove these temp variables altogether?

```
return !hasMultiRowGroups(leftBucketed, width) && !hasMultiRowGroups(rightOversized, width);
```

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  bool leftOversized = hasMultiRowGroups(leftBucketed, width);
+  bool rightOversized = hasMultiRowGroups(rightBucketed, width);
+
+  return !leftOversized || !rightOversized;
```
Should this be && ?

kagamiori replied:
If only one of the result sets satisfies the requirement of every group contains a single row, we could still use the sort-merge algorithm with epsilon comparator, because the extra rows in the group in the other result set would be mismatch anyway. 

E.g., expected result is [(1, 0.1, 1.1), (2, 2.2)] and actual result is [(1, 0.09999, 1.09999), (1, 0.10001, 1.100001)]. Sort-merge still works.

mbasmanova reviewed (commented):
@kagamiori Thank you for iterating on this PR. These refactors make the code much easier to read. I still find usage of term "bucket" confusing though. Some follow-up comments below.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+bool equalPrefix(
+    const MaterializedRow& currentRow,
+    const MaterializedRow& previousRow,
+    size_t width) {
```
nit: width -> n

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& sortedRows,
+    size_t width) {
```
width -> n

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+}
+
+// Return true if there are multiple rows with the same prefix of the first
+// width columns. The argument groupedRows should be sorted by the first
```
groupedRows -> sortedRows

Consider moving "The argument ..." sentence to @param.

```
Returns true if there are multiple rows with the same values in the first n columns.
@param sortedRows Rows sorted by the first n columns.
```

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+
+std::vector<velox::column_index_t> getFloatColumnIndices(const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right) {
+  std::vector<velox::column_index_t> floatColumnIndices;
```
floatColumnIndices -> indices or columnIndices (float is implied by the function name)

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  auto width = leftRow.size();
+  for (auto i = 0; i < width; ++i) {
+    if (leftRow[i].kind() != rightRow[i].kind()) {
+      VELOX_FAIL("Expect the same type between two result sets.");
```
use VELOX_CHECK_EQ(leftRow[i].kind(), rightRow[i].kind(), "...");

This doesn't matter here, but we should use Type::equivalent for checking whether types are the same.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  return floatColumnIndices;
+}
+
+// Sort rows in left and right into leftSorted and rightSorted
```
Sort -> Sorts

```
Sorts two lists of rows by non-floating point columns. Returns true if non-floating point columns contain unique values in both lists.
```

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// Sort rows in left and right into leftSorted and rightSorted
+// and returns whether the sorting by unique keys is successful. The sorting is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
```
I feel that this check doesn't belong to this function. This function should just assume that this is the case, no?

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+  }
+
+  auto floatColumnIndices = getFloatColumnIndices(left, right);
+  if (floatColumnIndices.empty()) {
```
This logic doesn't seem to be described in the function comment. I'm thinking that this function is doing too many things. It might be cleaner to let it just do sorting and take list of floating point columns as input.

I'm also wondering whether the logic of moving floating point columns to the end would result in confusing error messages where missing and extra rows will be printed in an "unexpected" order that doesn't match the order of columns in the query plan. Would it be cleaner to do the sorting without moving the columns?

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -731,6 +929,70 @@ static bool compareMaterializedRows(
+    const std::string& message) {
+  std::vector<MaterializedRow> actualReordered;
+  std::vector<MaterializedRow> expectedReordered;
+  bool haveUniqueKeys = sortByUniqueKey(
```
haveUniqueKeys - this is a nice variable name, but it is misleading as sortByUniqueKey return true also when there are no floating-point columns. It might be cleaner to first identify the floating-point columns and then call sortByUniqueKey only if there is at least one such column.

kagamiori replied:
This is a good point, along with other good suggestions about this function. I'll update it. Thanks!

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -731,6 +929,70 @@ static bool compareMaterializedRows(
+  bool haveUniqueKeys = sortByUniqueKey(
+      actualRows, expectedRows, actualReordered, expectedReordered);
+
+  bool match = false;
```
Is this variable used?

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.h:
```diff
@@ -185,6 +166,25 @@ velox::variant readSingleValue(
     const std::shared_ptr<const core::PlanNode>& plan,
     int32_t maxDrivers = 1);
 
+/// assertEqualResults() has slightly different behaviors depending on the
```
Perhaps, streamline a bit.

```
This method has limited support for results with floating-point columns. 
...
```

mbasmanova reviewed (commented):
@kagamiori Some follow-up comments.

mbasmanova commented on velox/vector/fuzzer/VectorFuzzer.h:
```diff
@@ -227,6 +227,10 @@ class VectorFuzzer {
   TypePtr randType(int maxDepth = 5);
   RowTypePtr randRowType(int maxDepth = 5);
 
+  // Generate a random non-floating-point primitive type to be used as join keys
```
typo: join keys -> a join key or a group-by key

nit: drop ", etc."

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -772,11 +1110,21 @@ static bool compareOrderedPartitions(
+          left.second, right.second)) {
+    return rowsForComparison.resultsMatch();
+  }
+  // Compare the results directly without epsilon. This may cause false alarm
```
Consider adding a note similar to kDuckDbTimestampWarning to the error message in case there are floating-point columns.

mbasmanova replied:
> Consider adding a note similar to kDuckDbTimestampWarning to the error message in case there are floating-point columns.

Curious what are your thoughts on this.

kagamiori replied:
> > Consider adding a note similar to kDuckDbTimestampWarning to the error message in case there are floating-point columns.
> 
> Curious what are your thoughts on this.
Sorry, I missed this. I'll add a note.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -772,11 +1110,21 @@ static bool compareOrderedPartitions(
+  MaterializedRowsForEpsilonComparison rowsForComparison{
+      floatColumnIndices, nonFloatColumnIndices};
+  if (!floatColumnIndices.empty() &&
+      rowsForComparison.prepareForEpsilonComparison(
```
It looks like a call to prepareForEpsilonComparison is always followed by a call to resultsMatch. I wonder if it makes sense to change the signature of prepareForEpsilonComparison to return a boolean and remove resultsMatch.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,52 +549,310 @@ std::string toString(const MaterializedRow& row) {
+std::string MaterializedRowsForEpsilonComparison::getUserFriendlyDiff() {
+  VELOX_CHECK(
+      initialized_,
+      "Object must be initialized via validForEpsilonComparison first.");
```
A reference to validForEpsilonComparison needs updating as this method doesn't exist.

mbasmanova reviewed (commented):
Some comments.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -549,52 +549,291 @@ std::string toString(const MaterializedRow& row) {
+    return customLessThan(
+        lhs,
+        rhs,
+        [](const variant& lhs, const variant& rhs) { return lhs == rhs; },
```
== and < operators for non-floating point values are the same as equalsWithEpsilon and lessThanWithEpsilon. Hence, we can always to xxxWithEpsilon. If we do that we won't need customLessThan that takes equals and lessThan lambdas.

kagamiori replied:
The comparison here applies to floating-point columns too. We need strict comparison to avoid undefined behavior of sorting (as required by std library).

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -731,6 +968,108 @@ static bool compareMaterializedRows(
+    const MaterializedRowMultiset& actualRows,
+    const std::string& message) {
+  if (expectedRows.empty() != actualRows.empty()) {
+    LOG(ERROR) << generateUserFriendlyDiff(expectedRows, actualRows) << message;
```
Why do we log here and not use EXPECT_TRUE(false) << ...error message...? Same question for other logging statements.

kagamiori replied:
It's because I'd like to test unmatched comparisons in the unit test. If we use EXPECT_TRUE(false) here, the unit test always fail.

mbasmanova replied:
But doesn't that mean that the actual test won't fail. Do we need to ensure that all callers assert on the result. For example, the following code will not fail the test on failure, only log an error. Is this intentional?

```
  assertEqualResults(
      expectedRows,
      actualRows,
      kDuckDbTimestampWarning + "\nDuckDB query: " + duckDbSql);
```

mbasmanova replied:
Maybe there is a way to use a test listener to "catch" test failures in a test.

https://google.github.io/googletest/advanced.html#propagating-fatal-failures

kagamiori replied:
> But doesn't that mean that the actual test won't fail. Do we need to ensure that all callers assert on the result. For example, the following code will not fail the test on failure, only log an error. Is this intentional?
> 
> ```
>   assertEqualResults(
>       expectedRows,
>       actualRows,
>       kDuckDbTimestampWarning + "\nDuckDB query: " + duckDbSql);
> ```

Good catch! There are two callsites of assertEqualResults() that should assert the returned value is true. I'll fix them.

mbasmanova reviewed (approved):
@kagamiori Looks good to me % a few comments and questions. Thank you for adding a test.
